PSA Update return code for non-existing key in various key operations

maulik-arm opened the pull request:
## Description
Update return code for non-existing key in various key operations
Fixes issue #4162

## Status
**READY**

## Requires Backporting
No, PSA only.

## Migrations
If there is any API change, what's the incentive and logic for it.

YES | NO

## Additional comments
Any additional information that could be of interest

bensze01 commented on library/psa_crypto_slot_management.c:
```diff
@@ -371,8 +371,10 @@ psa_status_t psa_validate_key_location( psa_key_lifetime_t lifetime,
+psa_status_t psa_validate_key_persistence( psa_key_lifetime_t lifetime,
+                                           mbedtls_svc_key_id_t key_id )
 {
+    psa_status_t status;
```
The CI is complaining about this declaration being unused if `MBEDTLS_PSA_CRYPTO_STORAGE_C` is undefined. Please move it inside the ifdef below.

bensze01 reviewed (approved):
Looks good to me.

bensze01 commented on library/psa_crypto.c:
```diff
@@ -3364,6 +3369,10 @@ psa_status_t psa_sign_hash( mbedtls_svc_key_id_t key,
     status = psa_get_and_lock_key_slot_with_policy( key, &slot,
                                                     PSA_KEY_USAGE_SIGN_HASH,
                                                     alg );
+    /* Return invalid handle as per PSA API 1.0.0 if trying to copy persistent
```
_Originally posted by @stevew817 in https://github.com/ARMmbed/mbedtls/pull/4189#discussion_r587568248_
> You're not trying to copy a key here...

bensze01 commented on library/psa_crypto_slot_management.h:
```diff
@@ -203,12 +203,16 @@ psa_status_t psa_validate_key_location( psa_key_lifetime_t lifetime,
+ * \param[in] key_id    The key identifier.
  *
  * \retval #PSA_SUCCESS
+ * \retval #PSA_ERROR_INVALID_HANDLE
```
_Originally posted by @Summer-ARM in https://github.com/ARMmbed/mbedtls/pull/4189#discussion_r587332617_
> sorry, do I miss something? didn't see this return value in the function implementation.

bensze01 replied:
@Summer-ARM It looks like [`psa_validate_key_id()`](https://github.com/ARMmbed/mbedtls/pull/4198/files#diff-620cd14c340be363ec51956f5e66ec5233e9dffa97169020f0c3a0fd8d138359R387) is returning `PSA_ERROR_INVALID_HANDLE` instead of `PSA_ERROR_INVALID_ARGUMENT`, as it's documentation would suggest.

ronald-cron-arm replied:
There is indeed an issue here but not related to the issue this PR is solving. Thus not addressed in this PR. It is now addressed in #4279.

bensze01 reviewed (changes requested):
Copied over review comments from #4189

ronald-cron-arm commented on library/psa_crypto_slot_management.h:
```diff
@@ -203,12 +203,16 @@ psa_status_t psa_validate_key_location( psa_key_lifetime_t lifetime,
  *
  * \retval #PSA_SUCCESS
+ * \retval #PSA_ERROR_INVALID_HANDLE
+ *         \p key_id is not a valid key identifier.
```
There is `psa_validate_key_id()` to validate key identifiers thus why do you introduce such validation in this function that validates the persistence of a key ?

ronald-cron-arm reviewed (changes requested):
In its current version, I think that there are still APIs that are returning PSA_ERROR_DOES_NOT_EXIST. The approach of this PR is to fix the return value just before they return. I think it would be better to fix the source of the PSA_ERROR_DOES_NOT_EXIST error code for APIs which seems to be only `psa_get_and_lock_key_slot()`. This implies some changes in PSA test suites and in the coding of psa_open/close_key() I think. I don't know how much time you can spend on this. Please let me know if you are ok to follow the suggested path of if you prefer me to do it.

maulik-arm commented:
Hello Ronald,

I have limited time this week as I need to prepare tf-m for your 2.26 release(development branch) and on training next week. If you do not need it soon, I will start looking into this, but I may not be as fast as you 🙂 .  So I leave it upto you.

Thank you
Best Regards,
Maulik

________________________________
From: Ronald Cron <notifications@github.com>
Sent: Monday, March 8, 2021 2:32 PM
To: ARMmbed/mbedtls <mbedtls@noreply.github.com>
Cc: Maulik Patel <Maulik.Patel@arm.com>; Author <author@noreply.github.com>
Subject: Re: [ARMmbed/mbedtls] PSA Update return code for non-existing key in various key operations (#4198)


@ronald-cron-arm requested changes on this pull request.

In its current version, I think that there are still APIs that are returning PSA_ERROR_DOES_NOT_EXIST. The approach of this PR is to fix the return value just before they return. I think it would be better to fix the source of the PSA_ERROR_DOES_NOT_EXIST error code for APIs which seems to be only psa_get_and_lock_key_slot(). This implies some changes in PSA test suites and in the coding of psa_open/close_key() I think. I don't know how much time you can spend on this. Please let me know if you are ok to follow the suggested path of if you prefer me to do it.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/ARMmbed/mbedtls/pull/4198#pullrequestreview-606353079>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ASCUHYX7JDSMYH44AHRQ3GDTCTNYBANCNFSM4YTGTBVA>.

IMPORTANT NOTICE: The contents of this email and any attachments are confidential and may also be privileged. If you are not the intended recipient, please notify the sender immediately and do not disclose the contents to any other person, use it for any purpose, or store or copy the information in any medium. Thank you.

ronald-cron-arm commented:
I started to investigate what is need to change in the PSA crypto implementation, in order for PSA APIs not to return PSA_ERROR_DOES_NOT_EXIST error code but for the deprecated psa_open_key() and thus address completely #4162. Please have a look to [this](https://github.com/ronald-cron-arm/mbedtls/tree/does-not-exist). I've started to update the unit tests accordingly and I think it seems to go in the right direction, not completed though.

Could you check that the changes in PSA crypto as done [here](https://github.com/ronald-cron-arm/mbedtls/tree/does-not-exist) fix the issues related to the PSA_ERROR_DOES_NOT_EXIST when running the compliance test suite ?

ronald-cron-arm reviewed (changes requested):
Thanks for the changes. The final result seems correct to me but I think the commit history should be reworked:
1) Remove the three first commits.
2) When fixing the return codes in the tests please do it in three commits: one with the fixes for psa_open_key, one with the fixes for psa_close_key and the last one for the other APIs.
Do that locally and then force push. In case you are not familiar with the force push: git push --force <your fork remote name> maulik-arm/fix-4162.

gilles-peskine-arm commented:
Please use `git push --force-with-lease` when you do a force-push, never `git push --force`. It seems that `push --force` sometimes causes a race condition with GitHub that causes it to close the PR and makes it difficult to reopen.

ronald-cron-arm commented:
I never had any issue with --force but if --force-with-lease is safer I will use that.

gilles-peskine-arm reviewed (changes requested):
* This needs a changelog entry. I'd say under “Bugfix” since this fixes compliance with the PSA crypto specification, though arguably it could be considered a “Default behavior change”.
* The documentation of error codes for `psa_open_key` needs to be updated: it now returns `PSA_ERROR_DOES_NOT_EXIST` for invalid identifiers and not `INVALID_HANDLE` (alternatively, fix it to keep the current return codes — but I won't insist on it since this is a deprecated function and a suboptimal error code is acceptable).

Other than that, looks good to me.

maulik-arm commented:
Hello Ronald, Giles. 
Thank you for comments. I have reworked the commit history and added the change log entry file.  Regarding the documentation, I see that psa_open_key already mentions return value PSA_ERROR_DOES_NOT_EXIST for invalid ID. Is there any other place that needs update? 
Thank you

gilles-peskine-arm commented on ChangeLog.d/fix_return_type_for_invalid_crypto_key.txt:
```diff
@@ -0,0 +1,4 @@
+Bugfix
+   * Fixes the issue #4162. Returns PSA_ERROR_INVALID_HANDLE instead of
+     PSA_ERROR_DOES_NOT_EXIST for some key operations, if invalid key is
+     passed, to comply with PSA Crypto 1.0.0.
```
This entry is missing some context to be comprehensible and has grammatical mistakes. Suggested text:

> PSA functions other than psa_open_key now return PSA_ERROR_INVALID_HANDLE rather than PSA_ERROR_DOES_NOT_EXIST for an invalid handle, bringing them in line with version 1.0.0 of the specification. Fix #4162.

gilles-peskine-arm replied:
Please ensure that `tests/scripts/check_files.py` passes (see CI).

ronald-cron-arm commented:
Brilliant. Thanks for the rework. The flow of changes is easy to follow and understand that way. The documentation of psa_open_key already lists PSA_ERROR_DOES_NOT_EXIST and not PSA_ERROR_INVALID_HANDLE (as defined in v1 beta 3 of the spec). I think it is rather the case that the documentation and the code were not aligned before but are aligned now. Thus nothing to change here I think. It seems the only things left to take care of are:
. Update the change log
. Make sure tests/scripts/check_files.py is passing (not passing yet on the CI but seems related to the change log thus if you change it it may be ok with just modifying the change log).

ronald-cron-arm reviewed (dismissed):
LGTM

maulik-arm commented:
Thank you for review Ronald. There is Jenkins Test failure in checks: PR-4198-merging-TLS-Testing (https://jenkins-mbedtls.oss.arm.com/blue/organizations/jenkins/mbed-tls-pr-merge/detail/PR-4198-merge/18/pipeline) showing here.. I tried enabling config MBEDTLS_MEMORY_BUFFER_ALLOC_C, MBEDTLS_MEMORY_BACKTRACE. and MBEDTLS_PLATFORM_MEMORY in config.h and it is passing tests locally on my ubuntu machine (taking really long though). Is there any cause of concern here?

ronald-cron-arm commented:
I think it is just a CI timeout thus nothing to worry on your side.

gilles-peskine-arm commented on ChangeLog.d/fix_return_type_for_invalid_crypto_key.txt:
```diff
@@ -0,0 +1,4 @@
+Bugfix
+   * Fixes the issue #4162. Return PSA_ERROR_INVALID_HANDLE instead of
```
The changelog entry still does not follow the style of other changelog entries and is not grammatically correct.

* “fixes <s>the</s> issue #4162”, and we normally write “Fixes #nnnn.” at the end.
* “Return<s>s</s>”
* “if **an** invalid key is passed”
* For other compliance issue, we've been using the language “bring X in line with version 1.0.0 of the [PSA Crypto API] specification.”

See my suggested text above: https://github.com/ARMmbed/mbedtls/pull/4198#discussion_r596251675

maulik-arm replied:
Hello Giles.  I have updated the changelog entry. It fulfills all the requirements mentioned in the section "How to update Changelog Entry" on https://github.com/ARMmbed/mbedtls/blob/development/ChangeLog.d/00README.md. However, since you have some specific requirements on wordings, I would suggest if you can update the check_files.py to include these additional checks as well! :)

gilles-peskine-arm replied:
Your update still doesn't address all my comments.

AI is not advanced enough to automatically enforce grammar and style, let alone relevance and completeness.

maulik-arm replied:
Hello Giles, I think my latest Bugfix statement is grammatically correct, and clearly establishes the context. May I know what needs updating? Thanks
